agency/test/services/credential.test.ts

```diff
@@ -1,8 +1,267 @@
+      { cred_ex_id: 'credential-exchange-id', success: true, error: undefined }
+    ]);
+  });
+
+  it('should return a valid response even if a credential is not succesffully issued', async () => {
+    const credentialService = app.service('issuer/credential');
+    const profileWithSchema = await app.service('issuer/model').get(profile._id);
+    const params = {
+      profile: profileWithSchema,
+      headers: context.params.headers
+    } as unknown as IssuerServiceParams;
+
+    jest.spyOn(sleep, 'deferServiceOnce')
+      .mockReset()
+      .mockImplementationOnce((id) => mockSuccess(params, id))
+      .mockImplementationOnce((id) => mockError(params, id));
+
+    const credentials = [
+      { ...credential, reservation_number: '000003' },
+      { ...credential, reservation_number: '000004' }
+    ];
+    await credentialService.create(credentials, params);
+    expect(params.credentials.results).toBeInstanceOf(Array);
+    expect(params.credentials.results).toEqual([
+      { cred_ex_id: 'credential-exchange-id', success: true, error: undefined },
+      { cred_ex_id: 'credential-exchange-id', success: false, error: undefined }
+    ]);
+  });
+
+  it('should return a valid response even if a credential is not succesffully issued', async () => {
+    const credentialService = app.service('issuer/credential');
+    const profileWithSchema = await app.service('issuer/model').get(profile._id);
+    const params = {
+      profile: profileWithSchema,
+      headers: context.params.headers
+    } as unknown as IssuerServiceParams;
+
+    jest.spyOn(sleep, 'deferServiceOnce')
+      .mockReset()
+      .mockImplementationOnce((id) => mockError(params, id))
+      .mockImplementationOnce((id) => mockError(params, id));
+
+    const credentials = [
+      { ...credential, reservation_number: '000005' },
+      { ...credential, reservation_number: '000006' }
+    ];
+    await credentialService.create(credentials, params);
+    expect(params.credentials.results).toBeInstanceOf(Array);
+    expect(params.credentials.results).toEqual([
+      { cred_ex_id: 'credential-exchange-id', success: false, error: undefined },
+      { cred_ex_id: 'credential-exchange-id', success: false, error: undefined }
+    ]);
+  });
+
+  it('should throw a bad request error when existing schema is not found', async () => {
+    const credentialService = app.service('issuer/credential');
+    const profileWithSchema = await app.service('issuer/model').get(profile._id);
+    const params = {
+      profile: profileWithSchema,
+      headers: context.params.headers
+    } as unknown as IssuerServiceParams;
+    await credentialService.create({ ...credential, schema_name: 'unknown-schema' }, params);
+    const result = params.credentials.results as any;
+    expect(result).toBeInstanceOf(Object);
+    expect(result.cred_ex_id).toBeFalsy();
+    expect(result.success).toBe(false);
+    expect(result.error).toBeTruthy();
```

Comment:
Could we check that an error is thrown instead, and check the error message matches what we expect? See: https://jestjs.io/docs/expect#tothrowerror

Reply:
The challenge with this is the actual implementation, we catch the error and resolve the promise such that that `Promise.all` higher up in the call stack resolves with all results (whether they were successful or not). Therefore, in this case I don't test whether a rejection happens but whether the object returns with an error. I think we can discuss the possibility of changing the implementation to `reject` instead, although this may have adverse effects in the response retuned from the API call.